### DISCUSSION TITLE
Eagerly fetch nested relations?

### DISCUSSION THREAD

**Discussion by OskarPersson** [SeaQL/sea-orm#649] [Ideas]
When fetching a list of my table `Foo` I know that I can use `find_with_related` to eagerly fetch each Foo objects related `Bar`, giving me the following in return: `Vec<(Foo, Vec<Bar>)>`.  This is perfect for when there is only one level of relation. But what if I also have a one-to-one relation from `Bar` named `Qux` .

What should I use if I want all three from the same query, i.e. `Vec<(Foo, Vec<(Bar, Qux)>)>` ?

**Comment by billy1624** [comment]
Hey @OskarPersson, two levels eager loading is not supported at the moment.

> What should I use if I want all three from the same query, i.e. `Vec<(Foo, Vec<(Bar, Qux)>)>` ?

I guess it would be `Vec<(Foo, Vec<(Bar, Vec<Qux>)>)>`. Btw... this is way too complex loll

**Comment by OskarPersson** [comment]
hmm okay. How do you mean with "way too complex"? I feel like this is a pretty standard setup/problem and is supported in, for example, Django with [`select_related`](https://docs.djangoproject.com/en/4.0/ref/models/querysets/#select-related)/[`prefetch_related`](https://docs.djangoproject.com/en/4.0/ref/models/querysets/#prefetch-related)

> … then a call to `Book.objects.select_related('author__hometown').get(id=4)` will cache the related Person and the related City:

**Reply by frederikhors** [reply]
@OskarPersson I'm in your situation. I just opened https://github.com/SeaQL/sea-orm/discussions/1139 and found out this discussion.

Have you found a solution?

**Comment by VaZark** [comment]
Some kind of recursive call with a base case that falls back to lazy loading/fetching reference for cyclic relations might be one way to go about it. But I can see how it can quickly escalate in complexity.

OTOH, running a subquery can be a workaround but that makes the application code messy quickly too.

**Reply by frederikhors** [reply]
@VaZark I'm in your situation. I just opened https://github.com/SeaQL/sea-orm/discussions/1139 and found out this discussion.

Have you found a solution?

**Comment by billy1624** [comment]
> hmm okay. How do you mean with "way too complex"? I feel like this is a pretty standard setup/problem and is supported in, for example, Django with [`select_related`](https://docs.djangoproject.com/en/4.0/ref/models/querysets/#select-related)/[`prefetch_related`](https://docs.djangoproject.com/en/4.0/ref/models/querysets/#prefetch-related)
> 
> > … then a call to `Book.objects.select_related('author__hometown').get(id=4)` will cache the related Person and the related City:

At this early stage of sea-orm, we want to keep things simple, so we only allow fetching nested relation of one level deep, i.e. `Vec<(Foo, Vec<Bar>)>`. Anything beyond is just repetitive of it, fetching nested relation of two level deep, i.e. `Vec<(Foo, Vec<(Bar, Vec<Qux>)>)>`

**Comment by shutton** [comment]
+1 for me

It's very common to have tables that exist primarily of a set of IDs referencing other tables, and you want to rebuild a view of the original data. E.g., a tables such as:

## events
| event_id | time | item_id | purchaser_id |
| - | - | - | - |
| 6 | 2022-02-22 14:02 | 5 | 4 |

## purchasers
| purchaser_id | name |
| - | - |
| 4 | Bob Smith |

## items
| item_id | item_name |
| - | - |
| 5 | deluxe widget |

If you were selecting by `event_id`, you'd likely immediately want access to the records from the `items` and `purchasers` table.  But, at present, you have to pick one of these to fit into a `SelectTwo<E, F>`, and then perform a second query to obtain the `purchaser` record.

Other ORMs I've worked with implement methods (in this example) on the Event model that allow fetching the related records (potentially referencing cached values).  E.g., while there would be an `Event.item_id` structure member, an `Event.item()` method would return the associated Item model.

**Reply by frederikhors** [reply]
@shutton I'm in your situation. I just opened https://github.com/SeaQL/sea-orm/discussions/1139 and found out this discussion.

Have you found a solution?

**Comment by billy1624** [comment]
Hey @shutton, thanks for the proposal!

Related discussion:
- #557

**Comment by tyt2y3** [comment]
I think the problem mentioned by @shutton (A->B+C) is different from OP's problem (A->B->C)
The hard part of the problem is it only works if the relations are strictly one-to-one. If they are one to many, we will start to duplicate lots of data on the one side.
Anyway, I think the best way to tackle this scenario is to do a custom select with a custom struct.
